### PR TITLE
Docs lint/build/upload/publish pipeline

### PR DESCRIPTION
**Pull request by paulbalaji**
#### Description

- `gdk-for-unity-docs` pipeline for docs PRs
- lints and uploads a preview of changes to docs-testing (link available on BK build)
- buttons to promote a docs change to a preview on staging and then on to production

image

- Requires `DocsLinter` changes: https://github.com/spatialos/gdk-for-unity-shared-ci/pull/43

#### Tests

- linting steps work
- build/upload/tagging steps work

#### Documentation

- this helps docs

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: .buildkite/docs.definition.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+description: GDK for Unity Docs Pipeline
+github:
+  branch_configuration: []
+  default_branch: docs-current
+  pull_request_branch_filter_configuration:
+    - "docs-current"
+    - "docs-next"
+    - "docs/*"
+teams:
+- name: Everyone
```
Comment (efarem) [inline .buildkite/docs.definition.yaml]: Do you want everyone to be able to trigger the builds?

File: .toolsharerc [hunk 2]
```diff
@@ -0,0 +1,5 @@
+pinned_tool:
+  - tool: "improbadoc"
+    version: "2.20.2"
```
Comment (efarem) [inline .toolsharerc]: 2.20.5 is being released, you'll need this if you want to use the `LinkTo` macro (for cross linking between repos)

File: ci/tag-docs.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+fi
+
+ENVIRONMENT="production"
+
+setup_improbadoc "${ENVIRONMENT}"
+
+improbadoc list \
+    "${DOCS_TYPE}" \
+    --target="${DOCS_TARGET}" \
+    --environment="${ENVIRONMENT}" \
```
Comment (efarem) [inline ci/tag-docs.sh]: Environment for improbadoc should always be prod but `target` should change. Reason being that `environment` controls which service to talk to, `testing` and `staging` might not be stable.

Comment (paulbalaji) [reply ci/tag-docs.sh]: in which case - can I omit the `environment` argument and trust it will always talk to prod?

Comment (efarem) [reply ci/tag-docs.sh]: Yep

File: .buildkite/docs.definition.yaml (hunk 1 above)
Comment (paulbalaji) [reply .buildkite/docs.definition.yaml]: it's now `READ_ONLY`

File: .buildkite/docs.steps.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    artifact_paths:
+      - "./improbadoc.tag.output.json"
+    env:
+      DOCS_TYPE: "unity"
+      DOCS_VERSION: "alpha"
+      DOCS_TARGET: "testing"
+
+  - block: "Promote to staging!"
+
+  - label: "Preview docs in staging"
```
Comment (jamiebrynes7) [inline .buildkite/docs.steps.yaml]: Do we need testing & staging?

File: ci/lint-docs.sh [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+echo "Running Docs Linter"
+trap annotations ERR
+docker run \
+    -v $(pwd)/logs:/var/logs \
+    local:gdk-docs-linter \
+        "/var/logs/warnings.txt" \
+        "/var/logs/errors.txt"
+
+annotations
```
Comment (jamiebrynes7) [inline ci/lint-docs.sh]: Nit: `push_annotations`?

File: ci/tag-docs.sh [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+# Tag a version of the docs.
+#
+# Environment variables as arguments:
+#   DOCS_TYPE: "unity"
+#   DOCS_VERSION: "alpha"
+#   DOCS_TARGET: "testing", "staging" or "production"
+
+DOCS_HASH="$(buildkite-agent meta-data get "docs-hash")"
+if [ -z "${DOCS_HASH}" ]; then
+    echo "Expected docs-hash meta-data to be set."
```
Comment (jamiebrynes7) [inline ci/tag-docs.sh]: Won't the `buildkite-agent` command will fail before we get here?

File: ci/tag-docs.sh [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    exit 1
+fi
+
+if [ "${DOCS_TARGET}" != "production" ]; then
+    DOCS_VERSION="preview-${DOCS_HASH}"
+fi
+
+if [[ "${BUILDKITE_BRANCH}" != "master" && "${DOCS_TARGET}" == "production" ]]; then
```
Comment (jamiebrynes7) [inline ci/tag-docs.sh]: Is this correct? Our docs don't live on `master`?

File: ci/tag-docs.sh (hunk 6 above)
Comment (paulbalaji) [reply ci/tag-docs.sh]: yes - removing

File: ci/bootstrap.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    rm -rf "${SHARED_CI_DIR}"
+fi
+
+mkdir "${SHARED_CI_DIR}"
+
+# Workaround for being unable to clone a specific commit with depth of 1.
+pushd "${SHARED_CI_DIR}"
+    git init
+    git remote add origin "${CLONE_URL}"
+    git fetch --depth 20 origin fix/dont-check-improbadoc-esque-links
```
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply ci/bootstrap.sh]: Need to change the pinned branch :)

File: ci/lint-docs.sh [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+set -e -u -x -o pipefail
+
+if [[ -z "$BUILDKITE" ]]; then
+  echo "This script is only to be run on Improbable CI."
+  exit 1
+fi
+
+cd "$(dirname "$0")/../"
+
+ci/bootstrap.sh
```
Comment (jamiebrynes7) [inline ci/lint-docs.sh]: Do we need this here? I think we run `bootstrap.sh` on every build during the `post-checkout` right?

Comment (paulbalaji) [reply ci/lint-docs.sh]: we haven't set up a post-checkout for the docs branch - didn't think it was necessary as this is the only step that requires bootstrap to have run

Comment (jamiebrynes7) [reply ci/lint-docs.sh]: THat's probably fine then

**Pull request merged**